### DISCUSSION TITLE
CircuitPython support for mpremote

### DISCUSSION THREAD

**Discussion by bablokb** [bablokb/cp-shell#2] [General]
As suggested in https://github.com/adafruit/circuitpython/issues/9399#issuecomment-2208078274, we can discuss the issue of mpremote support for CP here.

When I tried to port rshell to CP, it failed for a number of reasons:

- sending python code to the device that is not supported by CP
- waiting for the string 'soft reboot' from the repl
- transferring data in chunks of 256 every 10ms
- probably some more minor reasons

The first issue will be hard to deal with in mpremote without inflating the code (e.g. CP does not have machine). 

Second issue: CP is localized, so instead of 'soft reboot' the repl might print 'ソフトリブート'. So mpremote would have to keep a dict of the translations used in CP and would also have to add code to guess (host locale) or overwrite it with a command line switch.

The third issue is with the code in https://github.com/micropython/micropython/blob/92484d8822635c125a7648d5b056ff7f78d62ab3/tools/mpremote/mpremote/transport_serial.py#L259

This code hard-codes chunks of 256 bytes every 10ms. In my tests with various (old) ESP32 boards I have to use 64 bytes every 500ms. So again, this would have to be changed by mpremote (I added two command line options for this purpose to cpshell).

**Comment by mattytrentini** [comment]
That's a fantastic summary, concisely highlighting the issues, thanks!

I'll try and dig in a little more to some of them; in particular I'd like to resolve the soft reboot issue. I also think it will be possible to determine which flavour of Python is running and use appropriate commands for that platform. I'll dig a little deeper...

Do you have any insight as to why such small chunks are required on CP for ESP32 boards? That will _dramatically_ impact performance!

**Reply by bablokb** [reply]
It is probably not necessary to really wait for the "soft reboot" message. The editor at code.circuitpython.org does not wait for it either (they have a similar implementation in JS). This is something to test later. After soft reboot you will get a prompt, and it is probably enough to wait for the prompt.

On my todo-list is to port the raw-paste mode from mpremote to cpshell. The code I linked to above only runs if raw paste mode is not available. With raw-paste mode, things should be much faster. But on my ESP32 boards the raw-paste mode has problems when I use it manually (i.e. directly from a terminal emulator), it just swallows arbitrary characters. So I am not sure this will help at all. 

Determining the flavor of Python is simple btw: just do an `import machine` and see if there is an exception or not.

**Comment by mattytrentini** [comment]
A quick update: I've been discussing the benefits of updating `mpremote` to support CircuitPython use with Damien; he's _definitely_ open to accepting such fixes. We just need to ensure they don't break MicroPython use.

First up though is to get [#11777](https://github.com/micropython/micropython/pull/11777) merged as it tidies up quite a lot of foundational features. 

Anyway, just keeping you posted!

**Reply by bablokb** [reply]
Great, thanks!
